clients/java/src/test/java/io/zeebe/client/impl/worker/JobWorkerImplTest.java

```diff
@@ -0,0 +1,238 @@
+        .pollInterval(Duration.ofMillis(10))
+        .atMost(Duration.ofSeconds(1))
+        .until(() -> gateway.getCountedPolls() > 3);
+    gateway.stopMeasuring();
+
+    // then polling is fast
+    assertThat(gateway.getTimeBetweenLatestPolls()).isLessThan(SLOW_POLL_THRESHOLD);
+
+    // when the gateway responds with errors
+    gateway.respondWith(new StatusRuntimeException(Status.RESOURCE_EXHAUSTED));
+
+    // then polling is slowed down
+    gateway.startMeasuring();
+    Awaitility.await()
+        .pollInterval(Duration.ofMillis(100))
+        .atMost(Duration.ofSeconds(5))
+        .untilAsserted(
+            () ->
+                assertThat(gateway.getTimeBetweenLatestPolls()).isGreaterThan(SLOW_POLL_THRESHOLD));
+  }
+
+  /**
+   * This mocked gateway is able to record metrics on polling for new jobs and easily switch how it
+   * responds to polling.
+   *
+   * <ul>
+   *   Due to the concurrent nature of the test setup and the job worker, 2 lock objects are used:
+   *   <li>responsesLock to lock access to the mocking of responses objects for test setup;
+   *   <li>metricsLock to lock access to the polling metrics objects.
+   * </ul>
+   */
+  private static final class MockedGateway extends GatewayImplBase {
+
+    // mocking of poll responses
+    private final Object responsesLock = new Object();
+    private boolean isInErrorMode = false;
+    private ActivateJobsResponse pollSuccessResponse = ActivateJobsResponse.newBuilder().build();
+    private StatusRuntimeException pollErrorResponse = new StatusRuntimeException(Status.UNKNOWN);
+
+    // polling metrics
+    private final Object metricsLock = new Object();
+    private boolean isMeasuring = false;
+    private long countedPolls = 0;
+    private Instant lastPoll = null;
+    private Duration timeBetweenLatestPolls = null;
+
+    @Override
+    public void activateJobs(
+        final ActivateJobsRequest request,
+        final StreamObserver<ActivateJobsResponse> responseObserver) {
+      synchronized (metricsLock) {
+        if (isMeasuring) {
+          final Instant now = Instant.now();
+          countedPolls++;
+          if (lastPoll != null) {
+            timeBetweenLatestPolls = Duration.between(lastPoll, now);
+          }
+          lastPoll = now;
+        }
+      }
+      synchronized (responsesLock) {
+        if (isInErrorMode) {
+          responseObserver.onError(pollErrorResponse);
+        } else {
+          responseObserver.onNext(pollSuccessResponse);
+          responseObserver.onCompleted();
+        }
+      }
+    }
+
+    public void respondWith(final List<ActivatedJob> jobs) {
+      synchronized (responsesLock) {
+        System.out.println("Now responding with jobs");
+        isInErrorMode = false;
+        pollSuccessResponse = ActivateJobsResponse.newBuilder().addAllJobs(jobs).build();
+      }
+    }
+
+    public void respondWith(final StatusRuntimeException throwable) {
+      synchronized (responsesLock) {
```

Comment:
Can you explain a bit why we use different lock objects instead of the synchronized keyword to use the `MockedGateway` instance itself as the lock? Do we expect contention when calling these methods? (ignoring `activateJobs` which is an inherited method)

Reply:
The access to the fields needs to be synchronised. Not all fields are used in all methods. So it worked out to split into 2 locks, to remove unnecessary blocking. I'm not sure whether this really leads to improved performance, but it made sense to me since these are really just 2 responsibilities (collecting metrics and mocking behaviour). The separation became more clear for me when I started splitting them. It would probably be better to extract 1 of these responsibilities but I did not want to make the test code too complex. WTDY? Do these locks improve/reduce understandability?

Reply:
I was just wondering why we don't just do `synchronized {` since we don't care about contention here (I presume), or just make the methods (except the activateJobs one) synchronized themselves. I figured it'd probably be simpler, but I don't care that much, I just wanted to know your thought process :+1: